aidial_adapter_openai/embeddings/azure_ai_vision.py

```diff
@@ -0,0 +1,153 @@
+    collect_embedding_inputs,
+)
+from aidial_adapter_openai.dial_api.resource import AttachmentResource
+from aidial_adapter_openai.dial_api.storage import FileStorage
+from aidial_adapter_openai.utils.auth import OpenAICreds
+from aidial_adapter_openai.utils.http_client import get_http_client
+from aidial_adapter_openai.utils.resource import Resource
+
+# The latest Image Analysis API offers two models:
+# * version 2023-04-15 which supports text search in many languages,
+# * the legacy 2022-04-11 model which supports only English.
+_VERSION_PARAMS = {
+    "api-version": "2024-02-01",
+    "model-version": "2023-04-15",
+}
+
+
+def _get_auth_headers(creds: OpenAICreds) -> dict[str, str]:
+    if "api_key" in creds:
+        return {"Ocp-Apim-Subscription-Key": creds["api_key"]}
+
+    if "azure_ad_token" in creds:
+        return {"Authorization": f"Bearer {creds['azure_ad_token']}"}
+
+    raise ValueError("Invalid credentials")
+
+
+class VectorizeResponse(BaseModel):
+    class Config:
+        extra = "allow"
+
+    vector: List[float]
+
+
+async def embeddings(
+    creds: OpenAICreds,
+    deployment: str,
+    endpoint: str,
+    file_storage: FileStorage | None,
+    data: dict,
+) -> EmbeddingResponse:
+    input = EmbeddingsRequest.parse_obj(data)
+
+    async def on_text(text: str) -> str:
+        return text
+
+    async def on_attachment(attachment: Attachment) -> Resource:
+        return await AttachmentResource(attachment=attachment).download(
+            file_storage
+        )
+
+    inputs_iter: AsyncIterator[str | Resource] = collect_embedding_inputs(
+        input,
+        on_text=on_text,
+        on_attachment=on_attachment,
+    )
+
+    inputs: List[str | Resource] = [input async for input in inputs_iter]
+
+    async def _get_embedding(input: str | Resource) -> VectorizeResponse:
+        if isinstance(input, str):
+            return await _get_text_embedding(creds, endpoint, input)
+        elif isinstance(input, Resource):
+            return await _get_image_embedding(creds, endpoint, input)
+        else:
+            assert_never(input)
+
+    tasks: List[asyncio.Task[VectorizeResponse]] = [
+        asyncio.create_task(_get_embedding(input)) for input in inputs
+    ]
+
+    responses = await asyncio.gather(*tasks)
+    vectors = [
+        Embedding(embedding=r.vector, index=idx)
+        for idx, r in enumerate(responses)
+    ]
+
+    n = len(vectors)
+    usage = Usage(prompt_tokens=n, total_tokens=n)
+
+    return EmbeddingResponse(model=deployment, data=vectors, usage=usage)
+
+
+async def _get_image_embedding(
+    creds: OpenAICreds, endpoint: str, resource: Resource
+) -> VectorizeResponse:
+    resp = await get_http_client().post(
+        url=endpoint.rstrip("/") + "/computervision/retrieval:vectorizeImage",
+        content=resource.data,
+        headers={
+            **_get_auth_headers(creds),
+            "content-type": resource.type,
+        },
+        params=_VERSION_PARAMS,
```

Comment:
User definitely can pass `api_version` by himself, if some new fields will be introduced into api response
Maybe let's add this `_VERSION_PARAMS` as defaults, if nothing passed?

And if something is passed, we validate that `api-version > 2024-02-01` and `model-version > 2022-04-11`

Reply:
api-version of the `/embeddings` endpoint doesn't have anything to do with api-version of the Azure AI Vision service. It's apples and oranges.

It's a bad user experience if one needs to go to Azure docs and find out what are the correct API versions for this particular service.

What we could do is to configure the default api version in the env var: `AZURE_AI_VISION_API_VERSION`, as it's been done for `DALLE3_AZURE_API_VERSION`. But I doesn't see any sense to introduce it while we still have a **single API version**.